add particle_rugosity

ramayo223 opened the pull request:
Script and ReadMe for calculating the (BFDH predicted) particle rugosity of a crystal structure in the CSD or from a local crystal structure file.

Alex-AMC commented on scripts/particle_rugosity/particle_rugosity.py:
```diff
@@ -0,0 +1,68 @@
+crystal = args.input_x
+
+# run code
+particle_rugosity(crystal)
```
I'd recommend wrapping this in a 
```if __name__ == '__main__':```

Example 

```python
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("input_x")
    args = parser.parse_args()
    crystal = args.input_x

    # run code
    particle_rugosity(crystal)```

Alex-AMC replied:
^^^ this is so you can import the `particle_rugosity` function without running the whole script every time.

ramayo223 replied:
modified, thanks!

Alex-AMC commented on scripts/particle_rugosity/particle_rugosity.py:
```diff
@@ -0,0 +1,68 @@
+
+# ================================
+# argument parser
+import argparse
```
This should be imported at the top of the file for more clear declarations.

Alex-AMC commented on scripts/particle_rugosity/particle_rugosity.py:
```diff
@@ -0,0 +1,68 @@
+# requires a CCDC license, and loading of the CSD-python miniconda environment if run from Linux command line
+
+# import ccdc modules
+import ccdc
+from ccdc.io import CrystalReader
+from ccdc import particle
+from ccdc.particle import Surface
+from ccdc import morphology
+from ccdc.morphology import BFDHMorphology
```
```suggestion
from ccdc.io import CrystalReader
from ccdc.particle import Surface
from ccdc.morphology import BFDHMorphology
```

I'd recommend removing the unused imports

Alex-AMC commented on scripts/particle_rugosity/particle_rugosity.py:
```diff
@@ -0,0 +1,68 @@
+def particle_rugosity(crystal):
+# load structure, either (1) crystal structure file in working directory, or 
+# (2) CSD refcode
+    head, sep, tail = crystal.partition('.')
+    if sep == '.':
```
I'd recommend looking at https://docs.python.org/3/library/os.path.html for handling paths for files, to make it a bit more safe and handle edge cases.

ramayo223 replied:
Added lines to check if file, try/except, for better input handling

Alex-AMC commented on scripts/particle_rugosity/particle_rugosity.py:
```diff
@@ -0,0 +1,68 @@
+# run BFDH morphology to determine predicted particle surface area = hkl planes, and d-spacing of the planes
+    morphology = BFDHMorphology(xtal)
+    facets = morphology.facets
+    f = facets[0]
```
Is this still needed?

ramayo223 replied:
I'm not sure I understand what part of this could be removed

Alex-AMC replied:
oh sorry, I didn't mean to highlight so many lines there. `f = facet[0]`
I don't believe it's doing anything there 🤔

Alex-AMC commented on scripts/particle_rugosity/ReadMe.md:
```diff
@@ -0,0 +1,85 @@
+## Summary
+Simulates the particle shape of the given crystal structure using BFDH mophology 
+prediction, then calculates the particle rugosity by determining the rugosity 
+of the facets and calculating the weighted average (by facet area).
```
Could you please mention in here that it uses the lowest rugosity for a given hkl to make sure people are aware that the offset is iterated through?

Alex-AMC commented on scripts/particle_rugosity/ReadMe.md:
```diff
@@ -0,0 +1,85 @@
+takes a crystal structure or refcode as an argument and will use the CCDC BFDH 
+morphology module to simulate the predicted particle shape. The facet hkl planes are
+then obtained and the rugosity of these planes is calculated using the [CCDC rugosity function](https://downloads.ccdc.cam.ac.uk/documentation/API/descriptive_docs/particle.html) (including a scan of 
+different off-sets from the origin to identify the minimum rugosity of that crystal 
```
offsets*

Alex-AMC replied:
Ah you can ignore the above comment, I see you mention the offsets here :)

Alex-AMC commented on scripts/particle_rugosity/ReadMe.md:
```diff
@@ -0,0 +1,85 @@
+
+## Licensing Requirements
+
+- CSD-Core
```
This also requires CSD-Particle Licence

Alex-AMC commented on scripts/particle_rugosity/particle_rugosity.py:
```diff
@@ -0,0 +1,68 @@
+
+# sum weighted rugosity values and print total particle rugosity
+    tot_rug = sum(w_part_rug)
+    print(f'{crystal} particle rugosity: ',tot_rug)
```
I don't know if you want to round off the float here? just to give a nicer answer to copy and paste into wherever a person might use it 
`print(f'{crystal} particle rugosity: ',round(tot_rug,3))`

ramayo223 replied:
yes, that makes for a cleaner output

Alex-AMC commented on scripts/particle_rugosity/ReadMe.md:
```diff
@@ -0,0 +1,85 @@
+~~~
+>>> import sys
+>>> import subprocess
+>>> subprocess.call([sys.executable, 'particle_rugosity.py', 'AABHTZ'])
+The crystal structure AABHTZ has been loaded
+Calculating particle rugosity ...
+AABHTZ particle rugosity:  1.5073781143188552
```
how come you're using the subprocess call here? 
Could you not just import the script via the normal way of given the instruction is to put the script where the API is ?  
Such as 

```python 
from particle_rugosity import particle_rugosity 

particle_rugosity('AABHTZ')
``` 
Would that not work?

ramayo223 replied:
fixed this

Alex-AMC reviewed (changes requested):
Hey, thank you for adding this script to the repo. It looks great. 

There are a couple of things that could be improved in terms of making the script a little bit more clear. 

I tested the script on the latest 2023.1 version of the CSD and it ran fine 👍  
Something you might want to consider if you're going to be doing high throughput calculation is the efficiency of calculating every facet property.  

There is a way of reducing the computation significantly by identifying equivalent facets and only calculating the rugosity for each group.

ramayo223 commented on scripts/particle_rugosity/particle_rugosity.py:
```diff
@@ -0,0 +1,68 @@
+from ccdc import morphology
+from ccdc.morphology import BFDHMorphology
+
+def particle_rugosity(crystal):
```
I think I've resolved this, let me know if it can be made even more clear

ramayo223 commented:
Committed an update addressing most comments. I'm not sure what could be removed regarding the BFDH morphology facet determination.
Very true that determining symmetry-equivalent facets would reduce computation time for higher symmetry structures, something I'll look into adding at a later date.
Thank you for all the comments and suggestions!

Alex-AMC reviewed (approved):
Nice work 👍  Thanks for making all those changes :)
